### DISCUSSION TITLE
Power saving and interrupt

### DISCUSSION THREAD

**Discussion by PM04290** [SpenceKonde/megaTinyCore#1151] [Library discussion]
I changed my cards to go from ATtiny85 to ATtiny3216 thinking that in sleep mode I would have an identical consumption, except this is not the case when using the interrupt wake-up (4µA vs 450µA).
I have a very simple code:
deactivate all the inputs
```
for(uint8_t pin=0; pin<8; pin++) {
  (&PORTA.PIN0CTRL)[pin +  0] = PORT_ISC_INPUT_DISABLE_gc;
  (&PORTA.PIN0CTRL)[pin + 32] = PORT_ISC_INPUT_DISABLE_gc;
  (&PORTA.PIN0CTRL)[pin + 64] = PORT_ISC_INPUT_DISABLE_gc;
}
```
set the interrupt wake-up

```
attachInterrupt(PIN_PA5, onPinWakeup, FALLING);
```
go to sleep

```
set_sleep_mode(SLEEP_MODE_PWR_DOWN);
sleep_enable();
sleep_cpu();

```
I would like to point out that the Tiny has nothing connected, so there is no external consumption.

With this code i have **450µA**, if i comment attachInterrupt line, i have **under 1 µA**

Have you an idea to solve my problem, or is it a Tiny3216 bug

**Comment by hmeijdam** [comment]
This is no complete code, so who knows the problem is in the part you did not show.

**Comment by PM04290** [comment]
this is the complete code

```
#include "avr/sleep.h"
#define PIN_DEBUG_LED  PIN_PC2
volatile uint16_t sleepCounter = 0;

ISR(RTC_PIT_vect)
{
  RTC.PITINTFLAGS = RTC_PI_bm; // Clear interrupt flag by writing '1' (required)
  sleepCounter++;
}

void timerInit()
{
  while (RTC.STATUS > 0) {}                // Wait for all register to be synchronized
  RTC.CLKSEL      = RTC_CLKSEL_INT32K_gc;  // 32.768kHz Internal Ultra-Low-Power Oscillator (OSCULP32K)
  RTC.PITINTCTRL  = RTC_PI_bm;             // PIT Interrupt: enabled
  RTC.PITCTRLA    = RTC_PERIOD_CYC32768_gc //
                    | RTC_PITEN_bm;        // Enable PIT counter: enabled
}

void onPinWakeup()
{
  sleepCounter = 0x1FFF;
}

void setup() {
  pinMode(PIN_DEBUG_LED, OUTPUT);
  timerInit();
  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  sleep_enable();
}

void loop() {
  digitalWrite(PIN_DEBUG_LED, HIGH);
  delay(100);
  digitalWrite(PIN_DEBUG_LED, LOW);
  powerOff();
}

static inline void PIN_all_disable(void) {
  for (uint8_t pin = 0; pin < 8; pin++) {
    (&PORTA.PIN0CTRL)[pin] = PORT_ISC_INPUT_DISABLE_gc;
    (&PORTB.PIN0CTRL)[pin] = PORT_ISC_INPUT_DISABLE_gc;
#if defined(__AVR_ATtinyxy6__)
    (&PORTC.PIN0CTRL)[pin] = PORT_ISC_INPUT_DISABLE_gc;
#endif
  }
}


void powerOff()
{
  ADC0.CTRLA &= ~ADC_ENABLE_bm;
  PIN_all_disable();
  attachInterrupt(PIN_PA5, onPinWakeup, FALLING);
  sleepCounter = 0;
  do {
    sleep_cpu();
  } while (sleepCounter < 10);

  ADC0.CTRLA |= ADC_ENABLE_bm;
  pinMode(PIN_DEBUG_LED, OUTPUT);
}
```

**Comment by hmeijdam** [comment]
Some thoughts.

 Disabling all pins first and then attaching an interrupt to one of them and you observe that PA5 can still wakeup the MCU?

 Could PIN_PA5 be floating (nu pullup active) and therefore trigger the interrupt randomly, increasing power consumption?

I never used attachinterrupt() on these parts, but I notice that no interrupt flag is cleared in the onPinWakeup() ISR. does "attachinterrupt" take care of that? If not the ISR would keep firing.

**Comment by PM04290** [comment]
with or without INPUT_PULLUP, consumption is the same, this is not a full complete code, i just need to solve sleep consumption, not wakeup problems :-). 
what are you using to wakeup this kind of MCU by interrupt ?

**Comment by MX682X** [comment]
Wild guess, but you might be hit by an Errata. Many newer chips got their Errata updated with following Problem:

> 2.2.3 Write Operation Lost if Consecutive Writes to Specific Address Spaces
An ST/STD/STS instruction to address ≥ 64 followed by either an ST/STD instruction to address < 64
or a write to the SLPCTRL.CTRLA register will cause a loss of the last write.
Work Around
To avoid loss of write operation, use one of the following workarounds depending on address space:
• Insert an NOP instruction before writing to address < 64, or use the OUT instruction instead of ST/STD
• Insert an NOP instruction before writing to SLPCTRL.CTRLA register

even though the assembler files seems to be alright, the power consumption seems to correspond to the IDLE sleep.
Could you try to replace 
```
set_sleep_mode(SLEEP_MODE_PWR_DOWN);
sleep_enable();
```
with 
```
_NOP();
SLPCTRL.CTRLA = SLPCTRL_SMODE_PDOWN_gc | SLPCTRL_SEN_bm;
```
just to make sure?

**Comment by PM04290** [comment]
Same result. the problem is attachInterrupt(), not sleep mode setting.

**Comment by PM04290** [comment]
just try to replace attachInterrupt() by only

`PORTA.PIN5CTRL = PORT_ISC_FALLING_gc;
`
and consumption is the same.... too much

**Comment by PM04290** [comment]
I have found the problem; just a big hole in documentation (or exemple) ....

```
 // NOT GOOD
attachInterrupt(PIN_PA5, onInterrupt, FALLING) 
 // BETTER
attachInterrupt(digitalPinToInterrupt(PIN_PA5), onPinWakeup, FALLING); 

```

**Reply by hmeijdam** [reply]
Thanks for the feedback. Reminds me to continue to stay away from attachinterrupt() and work with the datasheet and the registers directly

onInterrupt  vs  onPinWakeup  ????  typo?
